proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1135 @@
+  internal var _count: Int
+}
+```
+
+It provides a buffer-like interface to the elements stored in that span of memory:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  public var count: Int { get }
+  public var isEmpty: Bool { get }
+
+  subscript(_ position: Int) -> Element { _read }
+}
+```
+
+Note that `Span` does _not_ conform to `Collection`. This is because `Collection`, as originally conceived and enshrined in existing source code, assumes pervasive copyability and escapability for itself as well as its elements. In particular a subsequence of a `Collection` is semantically a separate value from the instance it was derived from. In the case of `Span`, a sub-span representing a subrange of its elements _must_ have the same lifetime as the view from which it originates. Another proposal will consider collection-like protocols to accommodate different combinations of `~Copyable` and `~Escapable` for the collection and its elements.
+
+`Span`s representing subsets of consecutive elements can be extracted out of a larger `Span` with an API similar to the recently added `extracting()` functions of `UnsafeBufferPointer`:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  public func extracting(_ bounds: Range<Int>) -> Self
+}
+```
+
+The first element of a given span is _always_ at position zero, and its last it as position `count-1`.
+
+As a side-effect of not conforming to `Collection` or `Sequence`, `Span` is not directly supported by `for` loops at this time. It is, however, easy to use in a `for` loop via indexing:
+
+```swift
+for i in 0..<mySpan.count {
+  calculation(mySpan[i])
+}
+```
+
+#### Extensions to Standard Library and Foundation types
+
+```swift
+extension Array {
+  // note: this could borrow a temporary copy of the `Array`'s storage
+  var storage: Span<Element> { _read }
+}
+extension ArraySlice {
+  // note: this could borrow a temporary copy of the `ArraySlice`'s storage
+  var storage: Span<Element> { _read }
+}
+extension ContiguousArray {
+  var storage: Span<Element> { get }
+}
+
+extension Foundation.Data {
+  var storage: Span<UInt8> { get }
+}
+
+extension String.UTF8View {
+  // note: this could borrow a temporary copy of the `String`'s storage
+  var storage: Span<Unicode.UTF8.CodeUnit> { _read }
+}
+extension Substring.UTF8View {
+  // note: this could borrow a temporary copy of the `Substring`'s storage
+  var storage: Span<Unicode.UTF8.CodeUnit> { _read }
+}
+extension Character.UTF8View {
+  // note: this could borrow a temporary copy of the `Character`'s storage
+  var storage: Span<Unicode.UTF8.CodeUnit> { _read }
+}
+
+extension SIMD {
+  var storage: Span<Self.Scalar> { get }
+}
+extension KeyValuePairs {
+  var storage: Span<(Self.Key, Self.Value)> { get }
+}
+extension CollectionOfOne {
+  var storage: Span<Element> { get }
+}
+
+extension Slice: ContiguousStorage where Base: ContiguousStorage {
+  var storage: Span<Base.Element> { get }
+}
```

Comment:
The `ContiguousStorage` protocol has been moved to "Future directions", but it's still mentioned in the "Detailed design".

Reply:
Removed this.